godnsbl.go

```diff
@@ -97,91 +94,88 @@ func Blacklists() []string {
 	}
 }
 
-// RBLResults holds the results of the lookup.
-type RBLResults struct {
-	// List is the RBL that was searched
-	List string `json:"list"`
-	// Host is the host or IP that was passed (i.e. smtp.gmail.com)
-	Host string `json:"host"`
-	// Results is a slice of Results - one per IP address searched
-	Results []Result `json:"results"`
-}
-
-// Result holds the individual IP lookup results for each RBL search.
+// Result holds the individual result of an IP lookup for an RBL search.
 // nolint:maligned
 type Result struct {
-	// Address is the IP address that was searched
-	Address string `json:"address"`
 	// Listed indicates whether or not the IP was on the RBL
 	Listed bool `json:"listed"`
-	// RBL lists sometimes add extra information as a TXT record
-	// if any info is present, it will be stored here.
+	// RBL lists sometimes add extra information as a TXT record.
+	// If any info is present, it will be stored here.
 	Text string `json:"text"`
-	// Error represents any error that was encountered (DNS timeout, host not
-	// found, etc.) if any
-	Error bool `json:"error"`
-	// ErrorType is the type of error encountered if any
-	ErrorType error `json:"error_type"`
 }
 
 // Reverse the octets of a given IPv4 address
-// 64.233.171.108 becomes 108.171.233.64.
-func Reverse(ip net.IP) string {
-	if ip.To4() == nil {
-		return ""
-	}
-
-	splitAddress := strings.Split(ip.String(), ".")
-
-	for i, j := 0, len(splitAddress)-1; i < len(splitAddress)/2; i, j = i+1, j-1 {
-		splitAddress[i], splitAddress[j] = splitAddress[j], splitAddress[i]
+// '64.233.171.108' becomes '108.171.233.64'.
+// If it's not an v4 IP, returns nil.
+func Reverse(ip net.IP) net.IP {
+	ip4 := ip.To4()
+	if ip4 == nil {
+		return nil
 	}
 
-	return strings.Join(splitAddress, ".")
+	return net.IPv4(ip4[3], ip4[2], ip4[1], ip4[0])
 }
 
-func query(rbl string, host string, r *Result) {
-	r.Listed = false
-
-	lookup := fmt.Sprintf("%s.%s", host, rbl)
+func query(rbl string, ip net.IP) (*Result, error) {
+	lookup := fmt.Sprintf("%s.%s", Reverse(ip).String(), rbl)
 
 	res, err := net.LookupHost(lookup)
+	if err != nil && !recordNotFound(err) { // If the record doesn't exists, the IP isn't blacklisted
+		return nil, err
+	}
+
+	listed := false
+	text := ""
+
 	if len(res) > 0 {
-		r.Listed = true
-		txt, _ := net.LookupTXT(lookup)
+		listed = true
+
+		txt, err := net.LookupTXT(lookup)
+		if err != nil && !isNonFatalDNSError(err) {
```

Comment:
Previously errors here weren't handled at all.
I don't really like that. It's totally possible that this call fails for some random reason (timeout, doesn't exist, whatever), but it's also possible that there's actual a real permanent error. I don't want to ignore these possibilities, so I just did a special check to passthrough if we see these non fatal errors, but return if not.

Reply:
Nice catch. Actually, timeout often happens so that it is important to take care of it.